Several improvements for `Intervals<T>`

bjoeris opened the pull request:
## Description

Improvements:
- Iterators for Intervals<T> now behave more like std iterators:
  - `i->start()` syntax to access interval properties from the iterator (old version used `i.start()`)
  - added a `const_iterator` for `Intervals<T>`, which allows iteration over a `const Intervals<T>&`
- Renamed the interval bounds: `start` (inclusive lower bound) and `finish` (exclusive upper bound), to avoid confusion with `begin` and `end`, used for iterators.
- Removed the old `IntervalsIter::setValue` method, which tried to also handle splitting/merging intervals, and had confusing behaviour. This is was replaced by a simpler `setValue`, along with new `split` and `mergeLeft` methods.
- Added `update` and `merge` methods to `Intervals<T>`
- Added unit tests for `Intervals<T>`

baldurk commented on renderdoc/core/intervals.h:
```diff
@@ -168,4 +167,31 @@ struct Intervals
+template <typename T, typename Compose>
+void updateIntervals(Intervals<T> &a, uint64_t begin, uint64_t end, T val, Compose comp)
+{
+  for(auto i = a.find(begin); i != a.end() && i.start() < end; i++)
```
The comments for `find` above state that it finds the first interval that starts **after** the parameter. Is that definitely what we want here? In my mind this is basically taking an arbitrary range and then applying the compose function to all intervals that overlap it.

So if `begin` sits in the middle of an interval don't we want to conservatively over-estimate and apply the composition to the whole of that interval - even the part before `begin`, rather than under-estimate and apply to the next whole interval contained? Or do I misunderstand what the intention is here.

bjoeris replied:
`find(x)` "finds the interval containing `x`" (per the comment above the definition of `find`).

The comment you are referring to was meant to explain the implementation of `find` (find the interval after `x`, then step back one). Although now I'm trying to remember why I didn't just use `lower_bound` and skip the `it--`...

I will at least clean up that confusing comment, and see if the implementation of `find` can't also be simplified.

baldurk replied:
Ah, my mistake. Yes that could do with a bit of clarification.

Also I know I've been very lax with this and only have recently added them, but the `Intervals` class as well as these helpers are good candidates for writing some unit tests. See for example `basic_types_tests.cpp`.

bjoeris replied:
I did have some unit tests when I first wrote that class, but it looks like they were misplaced somewhere along the way. I will also add new tests for the helpers.

baldurk commented on renderdoc/core/intervals.h:
```diff
@@ -168,4 +167,31 @@ struct Intervals
+void mergeIntervals(Intervals<T> &a, Intervals<T> &b, Compose comp)
+{
+  auto bi = b.begin();
+  auto ai = a.find(bi.start());
```
I think the same as above applies here if the intervals don't match up exactly.

baldurk reviewed (changes requested):
I think this is fine but can you look at the use of `Intervals::find` I've highlighted? If I've misunderstood that's fine, but would be good to explain why in comments for future reference.

Commit:
Several improvements for `Intervals<T>`

Previously, members were accessed as methods of the iterator (e.g.
`iter.start()`); now this becomes `iter->start()`.

This also adds const iterators for `Intervals<T>`, which allows the use of
`const Intervals<T> &`.

The bounds for intervals have been renamed `start` (inclusive lower bound) and
`finish` (exclusive upper bound), to avoid confusion with `begin` and `end`,
used for iterators.

Finally, the behaviour of the old `setValue` method was very confusing, and this
has been replaced by three methods with behaviour that is much easier to
understad: `setValue`, `split` and `mergeLeft`.

Commit:
Add `update` and `merge` methods to `Intervals<T>`

Also updated build files and added tests for `Intervals<T>`.

bjoeris commented:
I made some larger changes to `Intervals<T>` that I think should substantially improve the clarity of the code here and code that uses `Intervals<T>`. Also added unit tests.

baldurk commented on renderdoc/core/intervals_tests.cpp:
```diff
@@ -0,0 +1,417 @@
+  return intervals;
+}
+
+void check_intervals(Intervals<uint64_t> &value, std::vector<Interval> expected)
```
Nit:
```suggestion
void check_intervals(Intervals<uint64_t> &value, const std::vector<Interval> &expected)
```

It doesn't really matter for test code but it's good to keep things consistent and pass by ref when we don't need a copy. There are a couple of other helper functions here taking `vector`s by value that could be changed.

bjoeris replied:
Done

baldurk commented on renderdoc/core/intervals_tests.cpp:
```diff
@@ -0,0 +1,417 @@
+    SECTION("update a sub-interval")
+    {
+      Intervals<uint64_t> test;
+      test.update(5, 10, 1, std::plus<uint64_t>());
```
I'd prefer using a lambda as you do below, or even a simple static helper function like `static uint64_t add(uint64_t a, uint64_t b)` over `std::plus`. I didn't intuit it at first glance, and it might be a portability issue since it was only added in C++14.

bjoeris replied:
Done

baldurk reviewed (changes requested):
This looks great now, thank you for the changes. I've highlighted a couple of minor points in the tests but otherwise I think it's ready to go.
